pandas/tests/frame/methods/test_compare.py

```diff
@@ -257,8 +258,17 @@ def test_compare_ea_and_np_dtype(val1, val2):
             ("b", "other"): np.nan,
         }
     )
-    result = df1.compare(df2, keep_shape=True)
-    tm.assert_frame_equal(result, expected)
+    if (
+        val1 is pd.NA
+        and parse_version(np.__version__) > parse_version("1.24.0")
+        and "dev" in np.__version__
+    ):
+        # can't compare with numpy array if it contains pd.NA
+        with pytest.raises(TypeError, match="boolean value of NA is ambiguous"):
+            result = df1.compare(df2, keep_shape=True)
+    else:
+        result = df1.compare(df2, keep_shape=True)
+        tm.assert_frame_equal(result, expected)
```

Comment:
@jbrockmendel @phofl ideas on what to do about this?

In the new numpy, comparisons with an array which contains `pd.NA` fail, see https://github.com/pandas-dev/pandas/issues/50124#issuecomment-1342958334

Reply:
pd.NA inside ndarray[object] is never going to work nicely.  my inclination is to explicitly say it isnt supported and say that raising in this situation is expected

Reply:
Is the main concern here that equality comparisons to an array with object dtype will return an object array instead of a boolean?

Reply:
> will return an object array

I may have misunderstood what you're asking, but comparisons with arrays of object dtype which contain pd.NA will raise:
```python
In [1]: np.array([1, 3]) == np.array([4, pd.NA])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 np.array([1, 3]) == np.array([4, pd.NA])

File ~/pandas-dev/pandas/_libs/missing.pyx:413, in pandas._libs.missing.NAType.__bool__()
    411 
    412     def __bool__(self):
--> 413         raise TypeError("boolean value of NA is ambiguous")
    414 
    415     def __hash__(self):

TypeError: boolean value of NA is ambiguous
```

Reply:
specifically, this line

https://github.com/pandas-dev/pandas/blob/0ca49fcf1e5d6a870f99624272a7e9332ea27593/pandas/core/arrays/masked.py#L730

is where the comparison would happen

Reply:
I think we want consistency with scalar comparisons

```python
>>> import pandas as pd
>>> pd.NA == pd.NA
<NA>
>>> pd.NA == 3
<NA>
```

Not sure I would expect different behavior when contained in a numpy array

Reply:
Is this possible? Because we have
```python
In [1]: pd.NA.__eq__(3)
Out[1]: <NA>
```
and then I presume (but haven't checked) numpy calls `bool` on the result, and we end up with `NA.__bool__` being called, but that returns `TypeError`

https://github.com/pandas-dev/pandas/blob/d6608313e211be0a44608252a3a31cf5220963f4/pandas/_libs/missing.pyx#L412-L413

Reply:
shall we move the conversation about what should be done to the issue, to not have it too scattered? https://github.com/pandas-dev/pandas/issues/50124#issuecomment-1344798063